### PR TITLE
Fix: Generate random callback

### PR DESCRIPTION
**Pull request by waahm7**
*Description of changes:*
- Introduce a generate random callback which has user_data as well. We can't use this callback in Swift without user_data.
- Add shutdown callback for `exponential_backoff_strategy`.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: include/aws/io/retry_strategy.h [hunk 1]
```diff
@@ -108,8 +113,20 @@ struct aws_exponential_backoff_retry_options {
...
     uint64_t (*generate_random)(void);
+
+    /*
+     * By default this will be set to use aws_device_random. If you want something else, set it here.
+     */
+    generate_random_fn *generate_random_impl;
+    /**
+     * Optional user data for the generate random generate_random_impl.
+     */
+    void *generate_random_user_data;
```
Comment (graebm) [inline include/aws/io/retry_strategy.h:129]: I'd like to see the code that's using this.

I'm worried that ... now that we're passing a `void *` from another language to C. that other language is going to need to know when C is done using that `void *`. Like, do we need to add a shutdown callback too?

File: include/aws/io/retry_strategy.h [hunk 2]
```diff
@@ -108,8 +113,20 @@ struct aws_exponential_backoff_retry_options {
...
+
+    /** Deprecated. Use generate_random_impl instead
+     * By default this will be set to use aws_device_random. If you want something else, set it here.
+     * */
     uint64_t (*generate_random)(void);
+
+    /*
+     * By default this will be set to use aws_device_random. If you want something else, set it here.
+     */
+    generate_random_fn *generate_random_impl;
```
Comment (graebm) [inline include/aws/io/retry_strategy.h:125]: ah, the eternal "what do I call v2" question

File: include/aws/io/retry_strategy.h [hunk 3]
```diff
@@ -35,6 +35,11 @@ typedef void(aws_retry_strategy_on_retry_token_acquired_fn)(
  */
 typedef void(aws_retry_strategy_on_retry_ready_fn)(struct aws_retry_token *token, int error_code, void *user_data);
 
+/**
+ * Optional function to supply your own generate random implementation
+ */
+typedef uint64_t(generate_random_fn)(void *user_data);
```
[the remark this answers is not in the document]
Comment (waahm7) [reply include/aws/io/retry_strategy.h]: Thanks, done.

File: include/aws/io/retry_strategy.h (hunk 2 above)
Comment (waahm7) [reply include/aws/io/retry_strategy.h]: Any better ideas?

File: include/aws/io/retry_strategy.h (hunk 1 above)
Comment (waahm7) [reply include/aws/io/retry_strategy.h:129]: Thanks, that is a great catch. I have added a shutdown callback and relevant changes in Swift at https://github.com/awslabs/aws-crt-swift/pull/147/files

**Review by graebm — APPROVED** [review]

**Pull request merged**